Add user_data_path parameter to ec2_lc module

vboginskey opened the pull request:
The existing `user_data` parameter is limited by AWS to 16KB. Much larger userdata can be submitted if gzipped, but it is not possible to use lookup plugins to read binary data.

This PR adds a `user_data_path` parameter to the `ec2_lc` module, so that binary data can be read directly by the module.

vboginskey committed:
Add user_data_path parameter to ec2_lc module

gundalow commented on lib/ansible/modules/cloud/amazon/ec2_lc.py:
```diff
@@ -67,6 +67,10 @@
+  user_data_path:
+    description:
+      - Path to the file that contains userdata for the ec2 instances
+    required: false
```
Please add the following line directly below `required: false`
`version_added: "2.3"`

gundalow commented on lib/ansible/modules/cloud/amazon/ec2_lc.py:
```diff
@@ -277,6 +291,7 @@ def main():
             key_name=dict(type='str'),
             security_groups=dict(type='list'),
             user_data=dict(type='str'),
+            user_data_path=dict(type='str'),
```
Would `type='path'` be better here?

gundalow commented on lib/ansible/modules/cloud/amazon/ec2_lc.py:
```diff
@@ -67,6 +67,10 @@
     required: false
+  user_data_path:
+    description:
+      - Path to the file that contains userdata for the ec2 instances
```
`Path to the file that contains userdata for the ec2 instances. Multualive exclusive with I(user_data).`

And similar on the user_data `description:` block

The `I(...)` means italics, so it's clearer we are referring to a parameter.

gundalow reviewed (changes requested):
needs_revision

vboginskey committed:
Improve user_data_path parameter documentation

- Specify mutual exclusivity
- Specify `version_added`
- Change module parameter type to `path`

vboginskey commented:
Thanks for the review @gundalow, I've addressed your feedback.

vboginskey committed:
Use correct comparison for user_data result

gundalow reviewed (approved):
shipit

vboginskey commented:
bot_status

willthames commented:
!needs_revision

willthames commented:
shipit

vboginskey commented:
bot_status

vboginskey commented:
bot_broken

willthames commented:
bot_status

mattclay commented:
Removed bot broken label and comment to give the bot another chance at reviewing the PR.

vboginskey commented:
Not sure what I need to do here @mattclay.

> needs_revision: True

Yet: https://github.com/ansible/ansible/pull/20138#issuecomment-279599556

> community_shipits: 1

This PR has received two, although the bot may not be counting the review comment.

mattclay commented:
@gundalow According to jctanner your ship it comment didn't count because it was in a review comment instead of a normal comment.

gundalow commented:
shipit

ryansb commented:
!needs_revision

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_lc.py:
```diff
@@ -250,6 +263,8 @@ def create_launch_config(connection, module):
                 result['block_device_mappings'][-1]['ebs'] = dict(snapshot_id=bdm.ebs.snapshot_id, volume_size=bdm.ebs.volume_size)
 
+    if launch_configs[0].user_data != "":
+        result['user_data'] = "hidden" # Otherwise, we dump binary to the user's terminal
```
I'm not sure we should always hide user data. I think there should either be some detection of binary vs. text, or this should be optional, since the output wouldn't be displayed unless the user was in `-vvv` or higher.

willthames replied:
One `-v` is enough to get module results these days.

A reasonable tradeoff might be

```
if user_data_path:
    result['user_data'] = 'hidden'
```

as the compelling reason to use `user_data_path` over `user_data` with a lookup is to allow zipped input. More sophisticated options might be to check if the data is compressed, or just binary (see e.g. http://stackoverflow.com/a/7392391/3538079) and hide if so.

ryansb replied:
I'd be happy with that method (only show if it's provided inline, not if it's in a file). Pretty simple, and if users have it in a file they can always open it/unzip it/whatever.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_lc.py:
```diff
@@ -188,6 +194,13 @@ def create_launch_config(connection, module):
 
+    if user_data_path:
+        try:
+            user_data_file = open(user_data_path, 'r')
```
When reading files, please use a `with` block so the file will always be closed. This code will never close the filehandle.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_lc.py:
```diff
@@ -188,6 +194,13 @@ def create_launch_config(connection, module):
+            user_data_file = open(user_data_path, 'r')
+            user_data = user_data_file.read()
+        except IOError as e:
+            module.fail_json(msg=str(e))
```
Please include the traceback with the failure here, by adding `exception=traceback.format_exc()` to the `fail_json` call.

vboginskey committed:
Include traceback on error and use with block

vboginskey committed:
Only hide user data if provided in file

vboginskey commented:
ready_for_review

willthames commented:
shipit

vboginskey commented:
@ryansb I'd appreciate it if you could take another look.

ryansb commented:
Thanks @vboginskey
